Total overhaul and improvements

AlicanHasirci opened the pull request:
This PR is not mergable yet, there are still some cleaning up to do; opening it up just to align and bounce ideas 💡 

While I was looking for a good solution to #16 I've figured that the things that I implement are adding new layers of complexity to way we handle different texts,  therefore rather than slaping a duct tape to fix problems, I thought coming with a fresh approach would be better by methodically observing changes that may occur and their implications.

Here are some changes and design pillars with related context, please do let me know If you think I've missed something or the point lacks clarity.

### Some things to consider before digging in
* Native arrays: All native arrays that have the original mesh data are inside `TextTweenManager` and kept inside their own single respective array. This is due to way jobs and burst working more effieciently on consecutive data inside memory.
* When a text is added: we find the total vertex count and see if we have enough space inside the array, if not allocate the required size and move the previous array into new one and then append the mesh data to the end of our current.
* When a text is removed: Revert the text to original, remove the MeshData belonging to text and shift the current data to fill up the space by removal.
* When a text is changed: If the text size increases shift all data to right, if not shift to left. Update all the MeshData effect to reflect the change.
* CharData: I've changed the nature of CharData to reflect every vertex rather than every character inside a job. So right we have same amount of CharData as Vertices. Still considering the pros, may revert to how it used to be.

### New Classes
* `TextTweenManager`: Main manager class that holds data and handle modifications
* `MeshData`: Class responsible for tracking the TMP meshes data inside the main native arrays.
* `Snapshot`: Disposable object to schedule jobs and apply modifications

### Things that will be added:
* The whole offset (now named overlap which sounds more intuitive?) is still missing. I think about implementing it in a way that won't require recreation of CharData array.
* Haven't check how things work in PlayMode, hence the missing `Update` function.

wallstop commented:
If we're doing NativeArray size manipulation, would it make sense to use [NativeList](https://docs.unity3d.com/Packages/com.unity.collections@0.0/api/Unity.Collections.NativeList-1.html) instead, which can help keep track of size and resize automatically for us?

wallstop commented on Editor/TextDataManagerInspector.cs:
```diff
@@ -0,0 +1,89 @@
+
+        private readonly Lazy<MethodInfo> _add = new(
+            () =>
+                typeof(TextTweenManager).GetMethod(
```
Should these use internalsvisibleto instead of reflection?

AlicanHasirci replied:
It is definitely more reasonable, haven't given much tought and directly copied the old solution but internals make it lot better.

wallstop commented on Runtime/MeshData.cs:
```diff
@@ -0,0 +1,92 @@
+namespace TextTween
+{
+    [Serializable]
+    public class MeshData
```
If you want this serializable by Unity I think it might need a default constructor, since it can no longer be default constructed due to the single specified constructor? If this works as is, then nevermind!

AlicanHasirci replied:
It seems like Unity use something like an `Activator` to create new instances, therefore a lacking default constructor is not an issue when creating an new `MeshData` on Unity.

wallstop commented on Runtime/Snapshot.cs:
```diff
@@ -0,0 +1,48 @@
+        private NativeArray<float4> Colors { get; }
+        private NativeArray<CharData> Chars { get; }
+
+        public Snapshot(TextTweenManager handler)
```
Can leave as an optimization until later, but it might be worth it to maintain persistent versions of these to reduce the need for allocations on change. ie, instead of snapshot on change, keep all the memory pooled somewhere and then copy state -> pooled state, then work off of pooled state. TempJob allocators are faster than persistent allocators, but they're not free.

Hard to say if this is worth it / faster without benchmarks though.

AlicanHasirci replied:
You are totally right, but I was not sure if the copying operation would take a bigger toll than the TempJob allocation itself. Benchmark is need for decision making.

wallstop commented on Runtime/Utilities/NativeArrayUtility.cs:
```diff
@@ -0,0 +1,142 @@
+                NativeArray<TA> newArray = new(length, Allocator.Persistent);
+                JobHandle handle = Move(ref array, ref newArray, 0, 0, array.Length);
+                handle.Complete();
+                array.Dispose();
```
Might recommend a:

try 
{
// Work
}
finally 
{
    array.Dispose();
}

pattern to 100% ensure there are no leaks just in case the code / math is buggy in Move/job completion. From my understanding, try/finally just generates IL such that the stuff in "finally" is executed after whatever code, it doesn't add any runtime cost (https://stackoverflow.com/a/4107320)

AlicanHasirci replied:
Great advice!

wallstop commented:
This is awesome! 

Minor nits: put the usings inside of the namespace 👀 

Approach is much cleaner. With this, it's very clear what pieces of code are responsible for what data and you now have high level, powerful operations. We can then tweak the implementation details to see if one approach for <operation> is more performant than the other, without having to tweak tons of internal state and deal with all of the bugs that come out of it.

I love it! Great work.

wallstop commented on Editor/TextDataManagerInspector.cs:
```diff
@@ -0,0 +1,89 @@
+
+        private void OnEnable()
+        {
+            _textsProperty ??= serializedObject.FindProperty("Texts");
```
Just checking, I've had weird experience with OnEnable where, even when I select different objects, it doesn't get called as expected.

Minor concern here: You select one object, `_textsProperty` gets set, you select another object, `_textsProperty` maintains its old reference (old object)

I've ran into so many bugs like this that for custom editors, I keep a pointer to the currently selected thing, and then in `OnInspectorGUI`, check if it's changed and do rehydration logic then.

But if this works reliably, great! That's the only thing that matters.

AlicanHasirci replied:
Initially my approach was to do it on OnInspectorGUI then I got some interesting results. My whole assumption according to observation  is this, `OnEnable` is called in two different occasions:
* During assembly reload
* When you switch to another object therefore letting GC collect current inspector code instance and then comeback.
So in both cases OnEnable acts like a constructor callback and recreates a different custom inspector instance even though both objects you switch between has inspected type.
That being said, I am fairly certain that `_textProperty` is always null and would love to know if there is a case this is proven otherwise

wallstop replied:
Cool! If OnEnable instantiates a new version of this, then all should be good. However, if re-uses an existing reference, then `_textsProperty` will have issues. But this is super easy to test - make two things, inspect one + changes stuff, inspect the other + change stuff, see what happens

wallstop commented on Runtime/Modifiers/TransformModifier.cs:
```diff
@@ -114,19 +114,13 @@ float progress
-                    _vertices[vertexOffset + i] += offset;
-                }
+
+                _vertices[index] -= offset;
```
Might be cheaper to grab the value, operate on it locally, then put it back instead of doing get -> mutate -> set, get -> mutate -> set, get -> mutate -> set, but probably has very minimal performance implications.

wallstop commented on Runtime/Snapshot.cs:
```diff
@@ -0,0 +1,48 @@
+        public JobHandle Schedule(float progress, IEnumerable<CharModifier> modifiers)
+        {
+            JobHandle handle = new();
+            foreach (CharModifier modifier in modifiers)
```
I think this no longer checks for null / disabled modifiers, not sure if that's intended.

Could also pass in an IReadOnlyList<CharModifier> and walk it in a for loop to avoid allocating an enumerator

AlicanHasirci replied:
Missing modifier enable check is unintended, must have missed it during refactor 😅 
You are right on the IEnumerable, great catch!

wallstop commented on Runtime/TextTweenManager.cs:
```diff
@@ -0,0 +1,188 @@
+
+        public void Remove(TMP_Text text)
+        {
+            if (!_meshData.TryGetValue(text, out MeshData meshData))
```
Nit: I think you could simplify/replace this with a Remove call (Remove returns if it was successful as well as the value removed)

wallstop commented on Runtime/TextTweenManager.cs:
```diff
@@ -0,0 +1,188 @@
+            );
+        }
+
+        public void Dispose()
```
Should this implement `OnDestroy` and call `Dispose` in that path?

AlicanHasirci commented:
@wallstop There is one discovery that needs sharing by the way. Apparently TextMeshPro buffers and grows the vertex data in a manner that makes our whole approach a lot more useful.
Say you've created a TMP_Text with one character. TMP creates a mesh with 8 vertices. Since every character has 4 vertices, TMP creates a mesh data that would suffice for 2 character by default. 
When you exceed this limit and add 1 more character making the total char count 3, textmesh **doubles** the mesh size and grows the mesh to accomodate 4 characters in total. This behaviour is consistent where TMP always allocates more according to your needs hence creating an extra buffering. By using the `text.mesh.vertexCount` we leverage this behaviour most of the time do no allocations at all unless we exceed this limitation. If we do we grow our array size according to TMP.

Also one more thing I would like to add is a `BufferSize` property to assign a default size for arrays, for occasions where devs know the final text size in advance.

AlicanHasirci commented:
> This is awesome!
> 
> Minor nits: put the usings inside of the namespace 👀
> 
> Approach is much cleaner. With this, it's very clear what pieces of code are responsible for what data and you now have high level, powerful operations. We can then tweak the implementation details to see if one approach for _operation_ is more performant than the other, without having to tweak tons of internal state and deal with all of the bugs that come out of it.
> 
> I love it! Great work.

Thanks a lot for the encouragement, means a lot. Happy to know that I was not the only one having struggling with the previous version, this one feels a lot more cleaner.

P.S.: Blame Rider for putting the using directives outside namespace, I myself always put them inside namespace since your review meanwhile converting people who do otherwise, mostly by talking down on them making encouraging them to contemplate on their atrocious behaviour 🤣

wallstop commented:
> P.S.: Blame Rider for putting the using directives outside namespace, I myself always put them inside namespace since your review meanwhile converting people who do otherwise, mostly by talking down on them making encouraging them to contemplate on their atrocious behaviour 🤣

:D You can set Rider up to automatically do it for you! https://www.jetbrains.com/help/rider/Namespace_Imports_Style.html

wallstop commented:
> @wallstop There is one discovery that needs sharing by the way. Apparently TextMeshPro buffers and grows the vertex data in a manner that makes our whole approach a lot more useful. Say you've created a TMP_Text with one character. TMP creates a mesh with 8 vertices. Since every character has 4 vertices, TMP creates a mesh data that would suffice for 2 character by default. When you exceed this limit and add 1 more character making the total char count 3, textmesh **doubles** the mesh size and grows the mesh to accomodate 4 characters in total. This behaviour is consistent where TMP always allocates more according to your needs hence creating an extra buffering. By using the `text.mesh.vertexCount` we leverage this behaviour most of the time do no allocations at all unless we exceed this limitation. If we do we grow our array size according to TMP.
> 
> Also one more thing I would like to add is a `BufferSize` property to assign a default size for arrays, for occasions where devs know the final text size in advance.

This is a great find and great stuff, looking forward to the integration.

AlicanHasirci commented:
Hey again!

Here are some changes and improvements on the branch:
* I've moved all `using` directives under namespace.
* collected **ALL** native arrays under a new class called `MeshArray` main goal of this change is to move all arrays under a single class for ease of use and extensibility. 
* Removed `TempJob` allocations, in place, we are allocating two `MeshArray`'s one for `_original` mesh data and one for `_modified` to change and apply. This approach looks a lot more allocation friendly and intuitive.
* Fixed some small issues you've pointed out.

wallstop commented on Editor/TextDataManagerInspector.cs:
```diff
@@ -0,0 +1,71 @@
+        private void OnEnable()
+        {
+            _manager = (TextTweenManager)target;
+            _textsProperty ??= serializedObject.FindProperty("Texts");
```
Just checking, did you test this switching between various TextTweenManagers? Concern is that without a domain reload, OnEnable will be called again every context switch, but with the same instance, and this value will be valid and non-null, therefore maintaining a reference to the old value.

Suggestion:
```csharp
_textsProperty = serializedObject.FindProperty(nameof(TextTweenManager.Texts));
```

wallstop commented on Editor/TextDataManagerInspector.cs:
```diff
@@ -0,0 +1,71 @@
+            _textsProperty ??= serializedObject.FindProperty("Texts");
+            for (int i = 0; i < _textsProperty.arraySize; i++)
+            {
+                _previous.Add(
```
Suggestion: Clear list prior to array walk

wallstop commented on Editor/TextDataManagerInspector.cs:
```diff
@@ -0,0 +1,71 @@
+
+                foreach (TMP_Text o in remove)
+                {
+                    if (o == null)
```
Is this check necessary? Should we remove no matter what, if there is a diff?

wallstop commented on Runtime/MeshArray.cs:
```diff
@@ -0,0 +1,140 @@
+        public JobHandle Schedule(float progress, IReadOnlyList<CharModifier> modifiers)
+        {
+            JobHandle handle = new();
+            foreach (CharModifier modifier in modifiers)
```
Nit: The foreach in the mono runtime on an IReadOnlyList will call a boxed `GetEnumerator()`, generating (small) amounts of garbage. If you only have an IReadOnlyList, the most efficient way (garbage-free) to enumerate it is a counting for loop.

foreach on a literal array / list call the efficient value-typed `GetEnumerator()` and do not generate garbage. The latest .Net runtime is working on fixing this (see [here](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-10/runtime)), but it will be a long time before Unity adopts the current .Net runtimes.

wallstop commented on Runtime/MeshArray.cs:
```diff
@@ -0,0 +1,140 @@
+
+        public void Dispose()
+        {
+            _vertices.Dispose();
```
Should these do any checks for IsCreated? Is Dispose safe? Only concern is someone passes in a non-persistent allocator and the native pointers are already disposed, not sure what the behavior of calling Dispose() on these will be in that case.

wallstop commented on Runtime/TextTweenManager.cs:
```diff
@@ -0,0 +1,186 @@
+
+        private void OnDisable()
+        {
+            Dispose();
```
I think you want to remove the text change event here, concern is this getting disabled, text being updated and that triggering (potentially undesirable) events in this code.

Also recommend keeping track of whether or not the event was added with a bool flag just to be super sure that addition/removal of the event is necessary.

wallstop commented on Runtime/TextTweenManager.cs:
```diff
@@ -0,0 +1,186 @@
+
+        private void Update()
+        {
+            if (!Application.isPlaying || Mathf.Approximately(_progress, Progress))
```
Is this check necessary? What will call Update if the Application isn't playing?

wallstop commented on Runtime/Utilities/TextDataUtility.cs:
```diff
@@ -0,0 +1,59 @@
+            return text.mesh.vertexCount;
+        }
+
+        public static bool Contains(this IEnumerable<MeshData> collection, TMP_Text text)
```
Nit / perf: you could make this an extension for IReadOnlyList and do a counting for loop to not create a boxed enumerator (no garbage), but this is in a low-volume call path, so doesn't super matter.

wallstop commented on Tests/TextDataManagerTests.cs:
```diff
@@ -0,0 +1,115 @@
+//         {
+//             foreach (MockTextData mock in _manager.Texts)
+//             {
+//                 for (int i = 0; i < mock.Vertices.Count; i++)
```
Oooh looking forward to these existing!

wallstop commented:
So good, dude. Really excellent work. All of the comments I have are pretty minor, I'm happy with this if you are. Really excited for this to get into main and see what the performance is / be able to play around with some performance improvement ideas.

AlicanHasirci commented:
Guess I am done with the improvements for now. Main change here was to move `Overlap`(`Offset` on main) to it's own modifier as 'IntervalModifier' which is quite intuitive and gives more granularity in terms of animation.

One thing to consider is, since we are running over all mesh data(including ones which are buffer by TextMesh itself, if the modifier does not check the data integrity it may produce some `AABB` error by canvas due to `NaN` generated by modifier.

I am happy with where this repo evolved to and excited to see where it gets, thanks a lot for effort and help.

wallstop commented:
> Guess I am done with the improvements for now. Main change here was to move `Overlap`(`Offset` on main) to it's own modifier as 'IntervalModifier' which is quite intuitive and gives more granularity in terms of animation.
> 
> One thing to consider is, since we are running over all mesh data(including ones which are buffer by TextMesh itself, if the modifier does not check the data integrity it may produce some `AABB` error by canvas due to `NaN` generated by modifier.
> 
> I am happy with where this repo evolved to and excited to see where it gets, thanks a lot for effort and help.

Glad to help! Will keep contributing as/if I get any cool ideas/features that you think would be a good fit for the repo. Feel free to reach out in the future if you'd like to collab on any other projects.
